{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d7cd1b81",
   "metadata": {},
   "source": [
    "This script is run in a python 3.5 environment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2582ece8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0f6d4583",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "60dccd49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read and concatenate all maf files (with snvs)\n",
    "\n",
    "path_snv_mafs = \"/home/rania/Documents/Tina/WES/mafs/new_mafs/snvs/SNVs/\"\n",
    "path_cnv_vcfs = \"/home/rania/Documents/Tina/WES/single_samples/cnv_vcfs/\"\n",
    "\n",
    "\n",
    "snv_mafs = os.listdir(path_snv_mafs)\n",
    "cnv_vcfs = os.listdir(path_cnv_vcfs)\n",
    "\n",
    "\n",
    "# store all mafs in a list\n",
    "snv_mafs_list = list()\n",
    "\n",
    "\n",
    "# store all read CNV vcf files in a list\n",
    "#cnv_list = list()\n",
    "#for i in range(len(cnv_vcfs)):\n",
    "#    cnv_vcf = vcf.Reader(open(path_cnv_vcfs + cnv_vcfs[i], 'r'))\n",
    "#    cnv_list.append(cnv_vcf)\n",
    "\n",
    "\n",
    "for i in range(len(snv_mafs)):\n",
    "    sampleID = snv_mafs[i].split(\".\")[0]\n",
    "    snv_maf = path_snv_mafs + snv_mafs[i]\n",
    "    #print(snv_maf)\n",
    "    snv_maf_df = pd.read_csv(snv_maf, skiprows=0, header=1, sep = \"\\t\")   \n",
    "    \n",
    "    # insert new columns\n",
    "    snv_maf_df['copy_num_status'] = np.nan\n",
    "    snv_maf_df['minor_copy_num_status'] = np.nan\n",
    "    snv_maf_df['sv_type'] = np.nan\n",
    "    snv_maf_df['cellular_fraction'] = np.nan\n",
    "        \n",
    "    # loop through the rows using iterrows()\n",
    "    for index, row in snv_maf_df.iterrows():\n",
    "        chrom = row['Chromosome']\n",
    "        pos = row['Start_Position'] \n",
    "        #print(chrom,pos)\n",
    "        \n",
    "        #read the cnv vcf of the same sample\n",
    "        #print(path_cnv_vcfs + sampleID + \".vcf\")\n",
    "        cnv_vcf = vcf.Reader(open(path_cnv_vcfs + sampleID + \".vcf\", 'r')) \n",
    "        \n",
    "        for record_ in cnv_vcf:\n",
    "            chrom_ = record_.CHROM\n",
    "            pos_ = record_.POS\n",
    "            end_ = record_.INFO['END']\n",
    "            tcn_ = record_.INFO['TCN_EM']\n",
    "            lcn_ = record_.INFO['LCN_EM']\n",
    "            svtype_ = record_.INFO['SVTYPE']\n",
    "            cellular_frac_ = record_.INFO['CF_EM']\n",
    "            log_odds_ratio_segment_ = record_.INFO['MAF_R']\n",
    "            log_odds_ratio_segment_cluster_ = record_.INFO['MAF_R_CLUST']\n",
    "            #print(chrom_, pos_, end_, tcn_, svtype_)\n",
    "            \n",
    "            if chrom == chrom_ and pos >= pos_ and pos <= end_ :\n",
    "                #print(\"Variant in CNV\")\n",
    "                snv_maf_df.at[index,'copy_num_status'] = tcn_\n",
    "                snv_maf_df.at[index,'minor_copy_num_status'] = lcn_\n",
    "                snv_maf_df.loc[index,'sv_type'] = svtype_\n",
    "                snv_maf_df.at[index,'cellular_fraction'] = cellular_frac_\n",
    "                snv_maf_df.at[index,'log_odds_ratio_segment'] = log_odds_ratio_segment_\n",
    "                snv_maf_df.at[index,'log_odds_ratio_segment_cluster'] = log_odds_ratio_segment_cluster_                                                                                  \n",
    "                break\n",
    "            else:\n",
    "                snv_maf_df.at[index,'copy_num_status'] = 2\n",
    "                snv_maf_df.at[index,'minor_copy_num_status'] = 1\n",
    "                snv_maf_df.loc[index,'sv_type'] = 'normal_diploid'\n",
    "                snv_maf_df.at[index,'cellular_fraction'] = 1  # set to 1 for normal diploid region\n",
    "                snv_maf_df.at[index,'log_odds_ratio_segment'] = 0\n",
    "                snv_maf_df.at[index,'log_odds_ratio_segment_cluster'] = 0\n",
    "                \n",
    "        \n",
    "    snv_mafs_list.append(snv_maf_df)\n",
    "                \n",
    "                                  \n",
    "\n",
    "snv_mafs_combined = pd.concat(snv_mafs_list)\n",
    "\n",
    "snv_mafs_combined.to_csv(\"/home/rania/Documents/Tina/WES/clonal_subclonal/snv_mafs_combined.tsv\", \n",
    "              sep=\",\", \n",
    "              index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7ebb0b69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read and concatenate all maf files (with INDELS)\n",
    "\n",
    "path_indels_mafs = \"/home/rania/Documents/Tina/WES/mafs/new_mafs/indels/INDELS/\"\n",
    "path_cnv_vcfs = \"/home/rania/Documents/Tina/WES/single_samples/cnv_vcfs/\"\n",
    "\n",
    "\n",
    "indels_mafs = os.listdir(path_indels_mafs)\n",
    "cnv_vcfs = os.listdir(path_cnv_vcfs)\n",
    "\n",
    "\n",
    "# store all mafs in a list\n",
    "indels_mafs_list = list()\n",
    "\n",
    "\n",
    "for i in range(len(indels_mafs)):\n",
    "    sampleID = indels_mafs[i].split(\".\")[0]\n",
    "    indels_maf = path_indels_mafs + indels_mafs[i]\n",
    "    #print(indels_maf)\n",
    "    indels_maf_df = pd.read_csv(indels_maf, skiprows=0, header=1, sep = \"\\t\")   \n",
    "    \n",
    "    # insert new columns\n",
    "    indels_maf_df['copy_num_status'] = np.nan\n",
    "    indels_maf_df['minor_copy_num_status'] = np.nan\n",
    "    indels_maf_df['sv_type'] = np.nan\n",
    "    indels_maf_df['cellular_fraction'] = np.nan\n",
    "        \n",
    "    # loop through the rows using iterrows()\n",
    "    for index, row in indels_maf_df.iterrows():\n",
    "        chrom = row['Chromosome']\n",
    "        pos = row['Start_Position'] \n",
    "        #print(chrom,pos)\n",
    "        \n",
    "        #read the cnv vcf of the same sample\n",
    "        #print(path_cnv_vcfs + sampleID + \".vcf\")\n",
    "        cnv_vcf = vcf.Reader(open(path_cnv_vcfs + sampleID + \".vcf\", 'r')) \n",
    "        \n",
    "        for record_ in cnv_vcf:\n",
    "            chrom_ = record_.CHROM\n",
    "            pos_ = record_.POS\n",
    "            end_ = record_.INFO['END']\n",
    "            tcn_ = record_.INFO['TCN_EM']\n",
    "            lcn_ = record_.INFO['LCN_EM']\n",
    "            svtype_ = record_.INFO['SVTYPE']\n",
    "            cellular_frac_ = record_.INFO['CF_EM']\n",
    "            log_odds_ratio_segment_ = record_.INFO['MAF_R']\n",
    "            log_odds_ratio_segment_cluster_ = record_.INFO['MAF_R_CLUST']\n",
    "            #print(chrom_, pos_, end_, tcn_, svtype_)\n",
    "            \n",
    "            if chrom == chrom_ and pos >= pos_ and pos <= end_ :\n",
    "                #print(\"Variant in CNV\")\n",
    "                indels_maf_df.at[index,'copy_num_status'] = tcn_\n",
    "                indels_maf_df.at[index,'minor_copy_num_status'] = lcn_\n",
    "                indels_maf_df.loc[index,'sv_type'] = svtype_\n",
    "                indels_maf_df.at[index,'cellular_fraction'] = cellular_frac_\n",
    "                indels_maf_df.at[index,'log_odds_ratio_segment'] = log_odds_ratio_segment_\n",
    "                indels_maf_df.at[index,'log_odds_ratio_segment_cluster'] = log_odds_ratio_segment_cluster_                                                                                  \n",
    "                break\n",
    "            else:\n",
    "                indels_maf_df.at[index,'copy_num_status'] = 2\n",
    "                indels_maf_df.at[index,'minor_copy_num_status'] = 1\n",
    "                indels_maf_df.loc[index,'sv_type'] = 'normal_diploid'\n",
    "                indels_maf_df.at[index,'cellular_fraction'] = 1  # set to 1 for normal diploid region\n",
    "                indels_maf_df.at[index,'log_odds_ratio_segment'] = 0\n",
    "                indels_maf_df.at[index,'log_odds_ratio_segment_cluster'] = 0\n",
    "                \n",
    "        \n",
    "    indels_mafs_list.append(indels_maf_df)\n",
    "                \n",
    "                                  \n",
    "\n",
    "indels_mafs_combined = pd.concat(indels_mafs_list)\n",
    "\n",
    "indels_mafs_combined.to_csv(\"/home/rania/Documents/Tina/WES/clonal_subclonal/indels_mafs_combined.tsv\", \n",
    "              sep=\",\", \n",
    "              index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b6f850d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
